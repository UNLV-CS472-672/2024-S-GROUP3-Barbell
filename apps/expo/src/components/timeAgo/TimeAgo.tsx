import { View, Text } from "react-native"
import calculateTimeAgo from "~/utils/calculateTime"

export default function TimeAgo({ notif }: { notif: any }) {
  const timeAgo = calculateTimeAgo(notif.createdAt)
  // this is just further error checking/bug prevention. if there is something wrong with the timeAgo calculation then we want 
  // to just return an empty view component so we don't get any weird "NaN ago" outputs or weird spacing. This can occur if 
  // the notification object doesn't have a time / if the time zone is wrong and the date is in the future.
  return timeAgo == "" ? <View className="pt-2 pb-2" /> :
                         <Text className="text-right pt-2 pb-2" style={{color: "#CACACA", fontSize: 11}}>{timeAgo}</Text>
}