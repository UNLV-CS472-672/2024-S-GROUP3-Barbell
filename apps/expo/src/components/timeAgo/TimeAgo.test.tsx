import React from 'react'

import { render, screen } from '@testing-library/react-native'

import calculateTimeAgo from '~/utils/calculateTime'
import TimeAgo from './TimeAgo'

describe('TimeAgo', () => {
  /*  */
  test('renders correctly for different time intervals', () => {
    const now = new Date()
    const cases = [
      { createdAt: new Date(now.getTime() - 30 * 1000), expected: 'Just now' },
      { createdAt: new Date(now.getTime() - 2 * 60 * 1000), expected: '2 minutes ago' },
      { createdAt: new Date(now.getTime() - 3 * 60 * 60 * 1000), expected: '3 hours ago' },
      { createdAt: new Date(now.getTime() - 2 * 24 * 60 * 60 * 1000), expected: '2 days ago' },
      { createdAt: new Date(now.getTime() - 2 * 7 * 24 * 60 * 60 * 1000), expected: '2 weeks ago' },
    ]

    for (const { createdAt, expected } of cases) {
      render(<TimeAgo createdAt={createdAt} />)
      const timeAgoText = screen.getByTestId('timeAgoText')
      expect(timeAgoText).toHaveTextContent(expected)
    }
  })

  /*  */
  test('renders an empty view for invalid or future dates', () => {
    const invalidDate = new Date('invalid')
    const futureDate = new Date(Date.now() + 60 * 60 * 1000) // 1 hour in the future

    render(<TimeAgo createdAt={invalidDate} />)
    expect(screen.getByTestId('timeAgoBadDate')).toBeTruthy()

    render(<TimeAgo createdAt={futureDate} />)
    expect(screen.getByTestId('timeAgoBadDate')).toBeTruthy()
  })

  /*  */
  test('calculateTimeAgo helper function works correctly', () => {
    const now = new Date()
    const cases = [
      { timeSent: new Date(now.getTime() - 30 * 1000), expected: 'Just now' },
      { timeSent: new Date(now.getTime() - 2 * 60 * 1000), expected: '2 minutes ago' },
      { timeSent: new Date(now.getTime() - 3 * 60 * 60 * 1000), expected: '3 hours ago' },
      { timeSent: new Date(now.getTime() - 2 * 24 * 60 * 60 * 1000), expected: '2 days ago' },
      { timeSent: new Date(now.getTime() - 2 * 7 * 24 * 60 * 60 * 1000), expected: '2 weeks ago' },
    ]

    for (const { timeSent, expected } of cases) {
      expect(calculateTimeAgo(timeSent)).toBe(expected)
    }
  })

  it('renders correctly and matches snapshot', () => {
    const testDate = new Date()
    const { toJSON } = render(<TimeAgo createdAt={testDate} />)

    expect(toJSON()).toMatchSnapshot()
  })
})
