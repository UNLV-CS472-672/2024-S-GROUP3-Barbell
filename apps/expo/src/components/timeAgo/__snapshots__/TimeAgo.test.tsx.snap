// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimeAgo renders correctly and matches snapshot 1`] = `
<Text
  className="pb-2 pt-2 text-right"
  style={
    {
      "color": "#CACACA",
      "fontSize": 11,
    }
  }
  testID="timeAgoText"
>
  Just now
</Text>
`;
